File: mobile/lib/core/models/consultation.dart
/// Consultation model for video consultations
class Consultation {
  final String id;
  final String appointmentId;
  final String roomName;
  final String? roomUrl;
  final String status;
  final DateTime? scheduledStart;
  final DateTime? patientJoinedAt;
  final DateTime? doctorJoinedAt;
  final DateTime? startedAt;
  final DateTime? endedAt;
  final int? durationMinutes;
  final bool recordingConsent;
  final String? recordingUrl;
  final int? patientRating;
  final int? doctorRating;
  final String? connectionQuality;
  final DateTime createdAt;
  final DateTime updatedAt;

  Consultation({
    required this.id,
    required this.appointmentId,
    required this.roomName,
    this.roomUrl,
    required this.status,
    this.scheduledStart,
    this.patientJoinedAt,
    this.doctorJoinedAt,
    this.startedAt,
    this.endedAt,
    this.durationMinutes,
    required this.recordingConsent,
    this.recordingUrl,
    this.patientRating,
    this.doctorRating,
    this.connectionQuality,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Consultation.fromJson(Map<String, dynamic> json) {
    return Consultation(
      id: json['id'],
      appointmentId: json['appointment_id'],
      roomName: json['room_name'],
      roomUrl: json['room_url'],
      status: json['status'],
      scheduledStart: json['scheduled_start'] != null
          ? DateTime.parse(json['scheduled_start'])
          : null,
      patientJoinedAt: json['patient_joined_at'] != null
          ? DateTime.parse(json['patient_joined_at'])
          : null,
      doctorJoinedAt: json['doctor_joined_at'] != null
          ? DateTime.parse(json['doctor_joined_at'])
          : null,
      startedAt: json['started_at'] != null
          ? DateTime.parse(json['started_at'])
          : null,
      endedAt:
          json['ended_at'] != null ? DateTime.parse(json['ended_at']) : null,
      durationMinutes: json['duration_minutes'],
      recordingConsent: json['recording_consent'] ?? false,
      recordingUrl: json['recording_url'],
      patientRating: json['patient_rating'],
      doctorRating: json['doctor_rating'],
      connectionQuality: json['connection_quality'],
      createdAt: DateTime.parse(json['created_at']),
      updatedAt: DateTime.parse(json['updated_at']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'appointment_id': appointmentId,
      'room_name': roomName,
      'room_url': roomUrl,
      'status': status,
      'scheduled_start': scheduledStart?.toIso8601String(),
      'patient_joined_at': patientJoinedAt?.toIso8601String(),
      'doctor_joined_at': doctorJoinedAt?.toIso8601String(),
      'started_at': startedAt?.toIso8601String(),
      'ended_at': endedAt?.toIso8601String(),
      'duration_minutes': durationMinutes,
      'recording_consent': recordingConsent,
      'recording_url': recordingUrl,
      'patient_rating': patientRating,
      'doctor_rating': doctorRating,
      'connection_quality': connectionQuality,
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt.toIso8601String(),
    };
  }

  int? get waitTimeMinutes {
    if (patientJoinedAt != null && startedAt != null) {
      return startedAt!.difference(patientJoinedAt!).inMinutes;
    }
    return null;
  }

  bool get isWaiting => status == 'waiting';
  bool get isInProgress => status == 'in_progress';
  bool get isCompleted => status == 'completed';
}

/// Join room response model
class JoinRoomResponse {
  final String roomUrl;
  final String jwtToken;
  final String roomName;
  final String consultationId;
  final String role;

  JoinRoomResponse({
    required this.roomUrl,
    required this.jwtToken,
    required this.roomName,
    required this.consultationId,
    required this.role,
  });

  factory JoinRoomResponse.fromJson(Map<String, dynamic> json) {
    return JoinRoomResponse(
      roomUrl: json['room_url'],
      jwtToken: json['jwt_token'],
      roomName: json['room_name'],
      consultationId: json['consultation_id'],
      role: json['role'],
    );
  }
}

/// Waiting room status model
class WaitingRoomStatus {
  final String status;
  final int? position;
  final int? estimatedWaitMinutes;
  final DateTime? patientJoinedAt;

  WaitingRoomStatus({
    required this.status,
    this.position,
    this.estimatedWaitMinutes,
    this.patientJoinedAt,
  });

  factory WaitingRoomStatus.fromJson(Map<String, dynamic> json) {
    return WaitingRoomStatus(
      status: json['status'],
      position: json['position'],
      estimatedWaitMinutes: json['estimated_wait_minutes'],
      patientJoinedAt: json['patient_joined_at'] != null
          ? DateTime.parse(json['patient_joined_at'])
          : null,
    );
  }

  String get waitingMessage {
    if (position != null) {
      return position == 1
          ? "You're next!"
          : "You are #$position in the queue";
    }
    return "Waiting for doctor";
  }

  String get estimatedWaitMessage {
    if (estimatedWaitMinutes != null) {
      return estimatedWaitMinutes! < 60
          ? "~$estimatedWaitMinutes minutes"
          : "~${(estimatedWaitMinutes! / 60).ceil()} hour(s)";
    }
    return "Calculating...";
  }
}

/// Doctor queue item model
class DoctorQueueItem {
  final String consultationId;
  final String appointmentId;
  final String patientId;
  final String patientName;
  final DateTime patientJoinedAt;
  final int waitTimeMinutes;
  final String? chiefComplaint;
  final bool isEmergency;

  DoctorQueueItem({
    required this.consultationId,
    required this.appointmentId,
    required this.patientId,
    required this.patientName,
    required this.patientJoinedAt,
    required this.waitTimeMinutes,
    this.chiefComplaint,
    required this.isEmergency,
  });

  factory DoctorQueueItem.fromJson(Map<String, dynamic> json) {
    return DoctorQueueItem(
      consultationId: json['consultation_id'],
      appointmentId: json['appointment_id'],
      patientId: json['patient_id'],
      patientName: json['patient_name'],
      patientJoinedAt: DateTime.parse(json['patient_joined_at']),
      waitTimeMinutes: json['wait_time_minutes'],
      chiefComplaint: json['chief_complaint'],
      isEmergency: json['is_emergency'] ?? false,
    );
  }
}

/// Doctor queue response model
class DoctorQueueResponse {
  final String doctorId;
  final List<DoctorQueueItem> queue;
  final int totalWaiting;

  DoctorQueueResponse({
    required this.doctorId,
    required this.queue,
    required this.totalWaiting,
  });

  factory DoctorQueueResponse.fromJson(Map<String, dynamic> json) {
    return DoctorQueueResponse(
      doctorId: json['doctor_id'],
      queue: (json['queue'] as List)
          .map((item) => DoctorQueueItem.fromJson(item))
          .toList(),
      totalWaiting: json['total_waiting'],
    );
  }
}

File: backend/app/schemas/consultation.py
"""
Consultation schemas for telemedicine.
"""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field

from app.models.consultation import (
    ConnectionQuality,
    ConnectionType,
    ConsultationStatus,
    ParticipantRole,
)


class ConsultationBase(BaseModel):
    """Base consultation schema."""

    appointment_id: UUID


class ConsultationCreate(ConsultationBase):
    """Schema for creating a consultation."""

    pass


class JoinRoomResponse(BaseModel):
    """Response when joining a consultation room."""

    room_url: str
    jwt_token: str
    room_name: str
    consultation_id: UUID
    role: ParticipantRole


class WaitingRoomStatus(BaseModel):
    """Status of patient in waiting room."""

    status: ConsultationStatus
    position: int | None = Field(
        None,
        description="Position in queue (1 = next)",
    )
    estimated_wait_minutes: int | None = Field(
        None,
        description="Estimated wait time in minutes",
    )
    patient_joined_at: datetime | None = None


class WaitingRoomJoinResponse(BaseModel):
    """Response when patient joins waiting room."""

    consultation_id: UUID
    status: ConsultationStatus
    message: str = "You are in the waiting room"
    position: int | None = None
    estimated_wait_minutes: int | None = None


class RecordingConsentRequest(BaseModel):
    """Request to submit recording consent."""

    consent: bool = Field(
        ...,
        description="Whether user consents to recording",
    )
    consent_text: str | None = Field(
        None,
        description="Legal consent text shown to user",
    )


class RecordingConsentResponse(BaseModel):
    """Response after submitting consent."""

    consultation_id: UUID
    user_role: ParticipantRole
    consent_given: bool
    consent_timestamp: datetime
    can_start_recording: bool = Field(
        ...,
        description="Whether recording can start (requires both parties)",
    )


class RatingRequest(BaseModel):
    """Request to submit post-consultation rating."""

    rating: int = Field(
        ...,
        ge=1,
        le=5,
        description="Rating from 1 (poor) to 5 (excellent)",
    )
    feedback: str | None = Field(
        None,
        max_length=500,
        description="Optional feedback text",
    )


class ConsultationParticipantResponse(BaseModel):
    """Schema for consultation participant."""

    id: UUID
    user_id: UUID
    role: str
    joined_at: datetime | None
    left_at: datetime | None
    device_type: str | None
    connection_type: str | None
    session_duration_minutes: int | None

    model_config = {"from_attributes": True}


class ConsultationRecordingResponse(BaseModel):
    """Schema for consultation recording."""

    id: UUID
    consultation_id: UUID
    patient_consent_at: datetime | None
    doctor_consent_at: datetime | None
    file_path: str | None
    file_size_mb: int | None
    duration_seconds: int | None
    uploaded_to_emr: bool
    expires_at: datetime | None
    has_full_consent: bool

    model_config = {"from_attributes": True}


class ConsultationResponse(BaseModel):
    """Schema for consultation response."""

    id: UUID
    appointment_id: UUID
    room_name: str
    room_url: str | None
    status: str
    scheduled_start: datetime | None
    patient_joined_at: datetime | None
    doctor_joined_at: datetime | None
    started_at: datetime | None
    ended_at: datetime | None
    duration_minutes: int | None
    wait_time_minutes: int | None
    recording_consent: bool
    recording_url: str | None
    patient_rating: int | None
    doctor_rating: int | None
    connection_quality: str | None
    created_at: datetime
    updated_at: datetime

    # Nested relationships (optional)
    participants: list[ConsultationParticipantResponse] | None = None
    recording: ConsultationRecordingResponse | None = None

    model_config = {"from_attributes": True}


class ConsultationListResponse(BaseModel):
    """Schema for listing consultations."""

    id: UUID
    appointment_id: UUID
    status: str
    scheduled_start: datetime | None
    started_at: datetime | None
    ended_at: datetime | None
    duration_minutes: int | None
    patient_rating: int | None

    model_config = {"from_attributes": True}


class DoctorQueueItem(BaseModel):
    """Schema for a patient in doctor's waiting queue."""

    consultation_id: UUID
    appointment_id: UUID
    patient_id: UUID
    patient_name: str
    patient_joined_at: datetime
    wait_time_minutes: int
    chief_complaint: str | None
    is_emergency: bool = False


class DoctorQueueResponse(BaseModel):
    """Schema for doctor's waiting room queue."""

    doctor_id: UUID
    queue: list[DoctorQueueItem]
    total_waiting: int


class ConnectionQualityUpdate(BaseModel):
    """Schema for updating connection quality."""

    quality: ConnectionQuality
    device_type: str | None = None
    connection_type: ConnectionType | None = None


class ConsultationEndRequest(BaseModel):
    """Request to end a consultation."""

    notes: str | None = Field(
        None,
        max_length=1000,
        description="Optional notes about the consultation",
    )
    connection_quality: ConnectionQuality | None = None
